### PR TITLE
Fallback to stat64 on EOVERFLOW error

### PR DESCRIPTION
**Pull request by TFSThiagoBR98**
This change is [Reviewable](https://reviewable.io/reviews/koreader/crengine/459)

### CODE REVIEW

**Comment by NiLuJe** [comment]
Isn't there already a `#define fstat fstat64` in `<sys/stat.h>` when `__USE_LARGEFILE64` is defined?

(I was actually thinking of something slightly more awful, by not trusting  `__USE_LARGEFILE64` at all, and checking if `fstat64` is defined, and forward-declaring it (and the struct) manually if not, because its protoype is pretty much set in stone).

**Comment by poire-z** [comment]
> (I was actually thinking of something slightly more awful, by not trusting `__USE_LARGEFILE64` at all, and checking if `fstat64` is defined, and forward-declaring it (and the struct) manually if not, because its protoype is pretty much set in stone).

I quite like this option, feels the less risky of side effects.

**Comment by NiLuJe** [comment]
(But the sanity checking and overflow checking is a nice touch I'd probably have skipped myself, so, kudos ;)).

**Comment by NiLuJe** [comment]
And paired with @poire-z's `HAVE_STAT64` define, we could bake this behind an ifdef and avoid my earlier crappy conditional patching idea ;).

**Comment by TFSThiagoBR98** [comment]
> Isn't there already a #define fstat fstat64 in <sys/stat.h> when __USE_LARGEFILE64 is defined?

[No only if i define _FILE_OFFSET_BITS=64 it will do it](https://github.com/bminor/glibc/blob/ae37d06c7d127817ba43850f0f898b793d42aea7/io/sys/stat.h#L233)

If i enable only __USE_LARGEFILE64 it will make stat64 avaliable and not replace all io calls with 64 bits functions, see [here](https://github.com/bminor/glibc/blob/ae37d06c7d127817ba43850f0f898b793d42aea7/io/sys/stat.h#L240)

**Comment by TFSThiagoBR98** [comment]
Here the [difference](https://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html#index-_005fLARGEFILE64_005fSOURCE)

File: crengine/src/lvstream.cpp [hunk 1]
```diff
@@ -801,11 +802,33 @@ class LVFileMappedStream : public LVNamedStream
             return error();
         }
         struct stat stat;
-        if ( fstat( m_fd, &stat ) ) {
-            CRLog::error( "Cannot get file size for %s", fn8.c_str() );
+        if ( fstat( m_fd, &stat ) < 0 ) {
+#ifdef HAS_STAT64
+            if ((int)errno == EOVERFLOW) {
```
Comment (NiLuJe) [inline crengine/src/lvstream.cpp:807]: nit: why is there an `int` cast on `errno`?

**Review by NiLuJe — APPROVED** [review]
Besides the casting nit, this looks good, and should be fairly free of side-effects, thanks! ;).

File: crengine/src/lvstream.cpp (hunk 1 above)
Comment (TFSThiagoBR98) [reply crengine/src/lvstream.cpp]: fixed, thanks

**Comment by poire-z** [comment]
We have to wait a few days for/after the coming up KOReader release before bumping this and the base PR into frontend. So, we may as well keep it unmerged here until then, in case you decide to add/fix something.

**Comment by TFSThiagoBR98** [comment]
@poire-z Ok, thanks for informing

**Comment by poire-z** [comment]
CI check was waiting for approval to be run, so I just ran it.
Some complains - dunno how serious, but if you can fix them, so we don't get these errors in future CI check, thanks :)

```
Running clang-tidy on crengine/src/lvstream.cpp
2 errors generated.
Error while processing /home/runner/work/crengine/crengine/crengine/src/lvstream.cpp.
Found compiler error(s).
/home/runner/work/crengine/crengine/crengine/src/lvstream.cpp:1388:94: error: use of undeclared identifier 'errno' [clang-diagnostic-error]
            CRLog::error( "Cannot get file size for %s, errno=%d, msg=%s", fn8.c_str(), (int)errno, strerror( errno )  );
                                                                                             ^
/home/runner/work/crengine/crengine/crengine/src/lvstream.cpp:1388:111: error: use of undeclared identifier 'errno' [clang-diagnostic-error]
            CRLog::error( "Cannot get file size for %s, errno=%d, msg=%s", fn8.c_str(), (int)errno, strerror( errno )  );
```

**Comment by NiLuJe** [comment]
Oh, those casts are also extraneous (and might actually be what clang's complaining about...).

**Comment by poire-z** [comment]
> Oh, those casts are also extraneous (and might actually be what clang's complaining about...).

Nope, still these complains :/

**Comment by TFSThiagoBR98** [comment]
This should fix the clang error

**Comment by poire-z** [comment]
Yep, it does.
Letting @NiLuJe confirm that stat64 and errno are really linux-only thingies :)

**Comment by NiLuJe** [comment]
> Yep, it does. Letting @NiLuJe confirm that stat64 and errno are really linux-only thingies :)

[Nor necessarily](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/fstat-fstat32-fstat64-fstati64-fstat32i64-fstat64i32?view=msvc-170), but it's not as if we cared ;p.

(And I'm unfamiliar with how both mingx & msvc map those to saner symbol names).

**Comment by poire-z** [comment]
Ok... but/so do you approve this PR ? :)

**Comment by NiLuJe** [comment]
Oh, definitely, don't give a damn about win ;).

**Comment by poire-z** [comment]
Merged, you can now bump crengine into your koreader-base PR (with `git submodule update --remote thirdparty/kpvcrlib/crengine` and `git add thirdparty/kpvcrlib/crengine`).

**Comment by TFSThiagoBR98** [comment]
@poire-z Done

**Pull request merged**